stdlib/json.gr

```diff
@@ -0,0 +1,2261 @@
+ *
+ * Following examples have whitespaces and line breaks replaced with visible
+ * charactes for illustrative purposes.
+ *
+ * `NoIndentation`
+ * ```
+ * {↵
+ * "currency":·"€",↵
+ * "price":·99.9↵
+ * }
+ * ```
+ *
+ * `IndentWithTab`
+ * ```
+ * {↵
+ * →"currency":·"€",↵
+ * →"price":·99.9↵
+ * }
+ * ```
+ *
+ * `IndentWithSpaces(2)`
+ * ```
+ * {↵
+ * ··"currency":·"€",↵
+ * ··"price":·99.9↵
+ * }
+ * ```
+ *
+ * `IndentWithSpaces(4)`
+ * ```
+ * {↵
+ * ····"currency":·"€",↵
+ * ····"price":·99.9↵
+ * }
+ * ```
+ */
+provide enum IndentationFormat {
+  NoIndentation,
+  IndentWithTab,
+  IndentWithSpaces(Number),
+}
+
+/**
+ * Controls how arrays are printed in custom formatting.
+ *
+ * Following examples have whitespaces and line breaks replaced with visible
+ * charactes for illustrative purposes.
+ *
+ * `CompactArrayEntries`
+ * ```
+ * []
+ * ```
+ *
+ * ```
+ * [1]
+ * ```
+ *
+ * ```
+ * [1,2,3]
+ * ```
+ *
+ * `OneArrayEntryPerLine`
+ * ```
+ * []
+ * ```
+ *
+ * ```
+ * [↵
+ * ··1↵
+ * ]
+ * ```
+ *
+ * ```
+ * [↵
+ * ··1,↵
+ * ··2,↵
+ * ··3↵
+ * ]
+ * ```
+ */
+provide enum ArrayFormat {
+  CompactArrayEntries,
+  OneArrayEntryPerLine,
+}
+
+/**
+ * Controls how objects are printed in custom formatting.
+ *
+ * Following examples have whitespaces and line breaks replaced with visible
+ * charactes for illustrative purposes.
+ *
+ * `CompactObjectEntries`
+ * ```
+ * {}
+ * ```
+ *
+ * ```
+ * {"a":1}
+ * ```
+ *
+ * ```
+ * {"a":1,"b":2,"c":3}
+ * ```
+ *
+ * `OneObjectEntryPerLine`
+ * ```
+ * {}
+ * ```
+ *
+ * ```
+ * {↵
+ * ··"a":·1↵
+ * }
+ * ```
+ *
+ * ```
+ * {↵
+ * ··"a":·1,↵
+ * ··"b":·2,↵
+ * ··"c":·3↵
+ * }
+ * ```
+ */
+provide enum ObjectFormat {
+  CompactObjectEntries,
+  OneObjectEntryPerLine,
+}
+
+/**
+ * Controls line ending type in custom formatting.
+ */
+provide enum LineEnding {
+  NoLineEnding,
+  LineFeed,
+  CarriageReturnLineFeed,
+  CarriageReturn,
+}
+
+/**
+ * Allows fine grained control of formatting in JSON printing.
+ */
+provide record FormattingSettings {
+  indentation: IndentationFormat,
+  arrayFormat: ArrayFormat,
+  objectFormat: ObjectFormat,
+  lineEnding: LineEnding,
+  finishWithNewLine: Bool,
+  escapeAllControlPoints: Bool,
+  escapeHTMLUnsafeSequences: Bool,
+  escapeNonASCII: Bool,
+}
+
+/**
+ * Compact formatting that minimizes the size of resulting JSON at cost of not
+ * being easily human readable.
+ *
+ * Only performs minimal string escaping as required by the ECMA-404 standard,
+ * so the result needs to be treated as proper unicode and is not safe to be
+ * transported in ASCII encoding.
+ *
+ * The following example have whitespaces, line breaks and control points
+ * replaced with visible characters.
+ * ```
+ * {"currency":"€","price":99.9,"currencyDescription":"EURO␡","script·unembeddable":"You·cannot·pase·a·JSON·object·directly·into·a·<script>·tag·in·HTML·if·it·contains·</·with·unescaped·forward·slash"}
+ * ```
+ */
+provide let defaultCompactFormat = () =>
```

Comment:
The reason is a bit silly now that I come back to it and it's probably better to change into a simple expression.  I was concerned about the fact that grain eagerly evaluates all top level let bindings and so this would result in higher (heap) memory consumption and slower startup times for programs don't use them. I'm not updated on latest developments in Grain, like treeshaking, so it may not even be the case anymore.